gencred/cmd/gencred/main.go

```diff
@@ -174,21 +272,132 @@ func Main() {
-
-	config, err := kubeconfig.ClientConfig()
-	if err != nil {
-		util.PrintErrAndExit(err)
+	var c config
+	if len(o.config) > 0 {
+		// Load from config yaml file
+		body, err := ioutil.ReadFile(o.config)
+		if err != nil {
+			util.PrintErrAndExit(err)
+		}
+		if err := yaml.Unmarshal(body, &c); err != nil {
+			util.PrintErrAndExit(err)
+		}
+	} else {
+		c.Clusters = []*clusterConfig{
+			{
+				Context:            &o.context,
+				Name:               o.name,
+				WithCertificate:    o.certificate,
+				WithServiceAccount: o.serviceaccount,
+				DurationInSeconds:  o.duration,
+				Overwrite:          o.overwrite,
+				Output:             &o.output,
+			},
+		}
 	}
 
-	clientset, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		util.PrintErrAndExit(err)
+	if o.refreshInterval == 0 {
+		if err := runOnce(c); err != nil {
+			util.PrintErrAndExit(err)
+		}
+		return
 	}
 
-	if err = writeConfig(o, clientset); err != nil {
-		util.PrintErrAndExit(err)
+	defer interrupts.WaitForGracefulShutdown()
+	interrupts.Tick(func() {
+		runOnce(c)
+	}, func() time.Duration { return o.refreshInterval })
+}
+
+func runOnce(c config) error {
+	// Make sure process everyone before crying.
+	var errs []error
+	for _, cc := range c.Clusters {
+		if cc.GKEConnection != nil && cc.Context != nil {
+			errs = append(errs, errors.New("gke and context are mutually exclusive"))
+			continue
+		}
+		if cc.DurationInSeconds == 0 {
+			cc.DurationInSeconds = defaultDurationInSeconds
+		}
+		var clientset *kubernetes.Clientset
+		if cc.Context != nil {
+			loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+			configOverrides := &clientcmd.ConfigOverrides{CurrentContext: *cc.Context}
+			kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)
+
+			config, err := kubeconfig.ClientConfig()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			clientset, err = kubernetes.NewForConfig(config)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		} else {
+			gkeService, err := container.NewService(context.Background())
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			cluster, err := container.NewProjectsLocationsClustersService(gkeService).Get(*cc.GKEConnection).Do()
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			decodedClientCertificate, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClientCertificate)
+			if err != nil {
+				fmt.Println("decode client certificate error:", err)
+				continue
+			}
+			decodedClientKey, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClientKey)
+			if err != nil {
+				fmt.Println("decode client key error:", err)
+				continue
+			}
+			decodedClusterCaCertificate, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClusterCaCertificate)
+			if err != nil {
+				fmt.Println("decode cluster CA certificate error:", err)
+				continue
+			}
```

Comment:
Why do we print these errors rather than handling them with the others in the aggregate?

Reply:
it was bad copy/paste from somewhere else, fixed